### PR TITLE
Partially defined sequences

### PR DESCRIPTION
**Pull request by mdehoon**
Add support for partially defined sequences to `Seq` objects.

Such partially defined sequences are frequently encountered in alignment files, for example in this line from a MAF file:
```
s hg38.chr7     117512683 36 + 159345973 TTGAAAACCTGAATGTGAGAGTCAGTCAAGGATAGT
```
where 117512683 is the starting position of the sequence on chromosome 7, and 159345973 is the size of chromosome 7. This line defines the sequence contents for positions 117512683:117512719; the sequence contents elsewhere is undefined.

The PR allows such sequences to be defined using a dictionary with the starting position and sequence contents of the defined sequence segment:
```python
>>> from Bio.Seq import Seq
>>> s = Seq({117512683: "TTGAAAACCTGAATGTGAGAGTCAGTCAAGGATAGT"}, length=159345973)
```

Operations on partially defined sequences are supported as much as possible, for example
```python
>>> s
Seq({117512683: 'TTGAAAACCTGAATGTGAGAGTCAGTCAAGGATAGT'}, length=159345973)
>>> len(s)
159345973
>>> s[117512690:117512700]
Seq('CCTGAATGTG')  # fully defined sequence
>>> s[:10]
Seq(None, length=10)
>>> s.reverse_complement()
Seq({41833254: 'ACTATCCTTGACTGACTCTCACATTCAGGTTTTCAA'}, length=159345973)
>>> print(s)
...
Bio.Seq.UndefinedSequenceError: Sequence content is only partially defined
```

More than one defined sequence segment is allowed, to enable addition of sequences:
```python
>>> s = Seq({3: "ABCD", 10: "XYZ"}, length=20)
>>> s
Seq({3: 'ABCD', 10: 'XYZ'}, length=20)
>>> s + s
Seq({3: 'ABCD', 10: 'XYZ', 23: 'ABCD', 30: 'XYZ'}, length=40)
```

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by mdehoon** [comment]
Not sure why black is complaining; it passes fine locally.

**Comment by peterjc** [comment]
Probably the exact version of black, currently the CI and recommended pre-commit setup is pinned to use black 19.10b0

**Comment by peterjc** [comment]
It seems very complicated. My initial reaction is "Why?", and if you really need this functionality, why put it in the base ``Seq`` object when it seems like an overhead for all the existing sequence object usage in Biopython.

**Comment by mdehoon** [comment]
> My initial reaction is "Why?",

How else are we going to do this? If we use `AlignIO` to parse a MAF file, we get the following:
```python
>>> from Bio import AlignIO
>>> alignments = AlignIO.parse("ucsc_test.maf", "maf")
>>> alignment = next(alignments)
>>> alignment[0]
SeqRecord(seq=Seq('AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG'), id='hg16.chr7', name='hg16.chr7', description='', dbxrefs=[])
```
and the position information is lost.

> It seems very complicated.

Some of the complications can be removed after we get rid of the deprecated code. Also, if we let `_PartiallyDefinedSequenceData` inherit from a `dict`, instead of having a dictionary as an attribute, it should get a bit simpler.

>why put it in the base Seq object 

The current `Seq` object was designed exactly with this in mind. To the user, all `Seq` objects look the same. But internally, how the `Seq` object stores the sequence data can be different. Usually it's stored directly as `bytes`, sometimes it's lazy-loaded from a 2bit file, sometimes it's read from `BioSQL`, sometimes the sequence is unknown, and sometimes the sequence is known partially. This can change constantly during a run, without the user having to worry about it, as in this example:
```python
>>> from Bio.Seq import Seq
>>> from Bio import SeqIO
>>> records = SeqIO.parse("sequence.bigendian.2bit", "twobit")
>>> record = next(records)
>>> seq = record.seq  # seq data are _TwoBitSequenceData for lazy-loading
>>> seq = seq[:10]  # seq data are `bytes`
>>> unknown_seq = Seq(None, length=10)  # unknown_seq data is _UndefinedSequenceData
>>> seq = seq + u + seq # seq data is a _PartiallyDefinedSequenceData
>>> seq = seq[8::20]  # seq data is bytes
```

> when it seems like an overhead for all the existing sequence object usage in Biopython.

Most of the functionality is implemented using inheritance; at run time, this gives zero overhead for usual `Seq` objects. The exceptions are:
- in `__add__`, where the type of `self` and `other` is checked. With a bit of work, I believe that this can also be handled through inheritance;
- in `__repr__`, where we can remove the overhead by rearranging the code and using inheritance;
- in `complement` and `reverse_complement`, but in the deprecated code only.

If it helps, I can add an example for the parser for EMBOSS `water` output to show how partially defined sequences are used in practice.

**Comment by mdehoon** [comment]
> Probably the exact version of black, currently the CI and recommended pre-commit setup is pinned to use black 19.10b0

Thank you, that was indeed the issue. It's fixed now.

**Comment by peterjc** [comment]
I would concede that in other alignment parsers we've just used the truncated sequence (the aligned region), and either discarded or at best record of the start/end in the ``SeqRecord`` annotations. It was practical though as usually it is only the aligned sequence which is of interest (see also the entire ``Bio.SearchIO`` object model, CC @bow).

What you are exploring a conceptually different approach... which would implications on alignment output too.

That the default usage of ``Seq`` is hardly affected is good news.

**Comment by mdehoon** [comment]
> I would concede that in other alignment parsers we've just used the truncated sequence (the aligned region), and either discarded or at best record of the start/end in the SeqRecord annotations. It was practical though as usually it is only the aligned sequence which is of interest (see also the entire Bio.SearchIO object model, CC @bow).

Yes I agree that it was practical and it made sense for the cases considered at the time. Also the `Bio.SearchIO` model makes sense, at it simply stores what is represented in the file (with appropriate modifications to be consistent with Python conventions).

But thinking forward, for a general alignment format, this wouldn't work well for all cases. For example, how to represent the alignment to the genome of an mRNA with long introns? Or how to represent an alignment if the sequence is not available (as in a BED file)?

So I have moved to a model where the sequences and the alignment are stored separately. The sequences are what went into the alignment program as input; the alignment only stores how the coordinates in one sequence map to those in another. The coordinates are always defined with respect to the original input sequences.

Partially defined sequences are useful only if the alignment program output contains the aligned sequence only (for example, in output from EMBOSS `water`). For consistency, if the alignment coordinates show that the query sequence aligns to positions 10:20 in the target sequence, then `target_seq[10:20]` must return the aligned sequence. If the output file does not include the full target sequence, then a partially defined sequence with the appropriate starting position is needed.

Note that this allows you to put in the full target sequence afterwards, without having to modify the alignment coordinates, along these lines:

```python
>>> genome = SeqIO.parse("humangenome.2bit", "twobit")  # human genome, lazy-loaded
>>> alignments = Align.parse("mybedfile.bed", "bed")
>>> for alignment in alignments:
...     chromosome_name = alignment.target.id  # alignment.target is a SeqRecord containing an undefined Seq object
...     alignment.target = genome[chromosome_name]  # put in the correct sequence; coordinates are still valid
...     print(alignment[0])  # This will show the aligned portion of the target sequence
...
```

**Comment by peterjc** [comment]
Interesting. You've clearly been looking at some large scale alignments of a kind I have not worked with.

Would this model work well with the pairwise alignments in SAM/BAM/CRAM files? They are a moving target and I stopped working on it a while ago but had a pure Python SAM/BAM parser working on top of ``Bio.bgzf`` a while back. In particular IIRC this was prior to tweaks to the BAM format for very long reads (PacBio and Oxford Nanopore) where the CIGAR string overflowed the original field size. But similar idea with the mapping coordinates encoded in the CIGAR string separately from the read and reference sequence.

**Comment by mdehoon** [comment]
> Would this model work well with the pairwise alignments in SAM/BAM/CRAM files? 

Yes this model is consistent with SAM/BAM/CRAM files; the SEQ field in SAM/BAM/CRAM stores the original query sequence, and the CIGAR string together with the starting coordinate POS define the alignment coordinates. We already have a writer for SAM output in `Bio.Align`:
```python
>>> from Bio.Align import clustal
>>> alignments = clustal.AlignmentIterator("kalign.aln")
>>> alignment = next(alignments)
>>> print(alignment)
GCTGGGGATGGAGAGGGAACAGAGT-T
|||||||||||||||||||||||||-.
GCTGGGGATGGAGAGGGAACAGAGTAG

>>> alignment.score = 0
>>> format(alignment, "sam")
'AT3G20900\t0\tTest1seq\t1\t255\t25M1I1M\t*\t0\t0\tGCTGGGGATGGAGAGGGAACAGAGTAG\t*\tAS:i:0\n'
```

**Comment by peterjc** [comment]
Getting off topic, but having ``format(alignment, "cigar")`` return ``'25M1I1M'`` in that example would be nice.

File: Bio/Seq.py
Comment (peterjc) [inline Bio/Seq.py:333]: CodeCov is pointing out that the above new code of partially defined sequences does not get tested when a fragment is over 60 letters.

File: Bio/Seq.py
Comment (peterjc) [inline Bio/Seq.py:2921]: CodeCov is flagging that replacing with a different length substring is not tested. I think I understand the logic here, but an explicit test seems wise.

File: Bio/Seq.py
Comment (peterjc) [inline Bio/Seq.py:2940]: Adding a test for negative length failing should be easy.

**Comment by mdehoon** [comment]
> Getting off topic, but having `format(alignment, "cigar")` return `'25M1I1M'` in that example would be nice.

We could add that. But how would you use it?

**Review by peterjc — CHANGES_REQUESTED** [review]
I'm not convinced of the need yet, but won't block this on those grounds alone.

I would like the test coverage improved though - I flagged a few, but probably some of the other CodeCov notes are worth addressing.

**Comment by peterjc** [comment]
The CIGAR string is reasonably well understood and compact, so I can see that being useful for debugging alignments. That isn't a compelling use case in of itself though.

File: Bio/Seq.py
Comment (mdehoon) [reply Bio/Seq.py:333]: Added a test

File: Bio/Seq.py
Comment (mdehoon) [reply Bio/Seq.py:2921]: Added a test

File: Bio/Seq.py
Comment (mdehoon) [reply Bio/Seq.py:2940]: Added a test

**Comment by mdehoon** [comment]
@peterjc  I believe all the relevant points from codecov have now been addressed.

For a practical application, I'll add (in a separate PR) a test case for `Tests/Emboss/water2.txt`, containing this alignment:
```
79 CGTTTGAGT-CTGGGATG     95
   ||||||||| ||||||||
 1 CGTTTGAGTACTGGGATG     18
```
where I will use the partially defined sequence `Seq({78: "CGTTTGAGTCTGGGATG"}, length=95)`.

**Review by peterjc — APPROVED** [review]
Is it worth pinging a few of the MAF contributors? I'd still like to have @Bow's thoughts, but perhaps he is not available. Also @MarkusPiotrowski - have you had a chance to look at this? Your work with pairwise alignments in Biopython probably makes you as good if not better qualified than me to review it.

Otherwise I don't think I have anything further to add: My opinion remains this seems complicated, and I personally don't have a use case for it, which makes me wary. I accept that you have got motivating examples though.

I suggest waiting a week before merging to give people a chance to comment?

**Comment by mdehoon** [comment]
@peterjc  Thanks. Sure, I can wait a week before merging.

**Comment by mdehoon** [comment]
@peterjc  See #3695  for an example from EMBOSS `water` where a partially defined sequence is needed.

**Pull request merged**
